File: NeetCode/Heap/Median_Finder/median__finder.py
from typing import Optional, List

import heapq


class MedianFinder:

    def __init__(self):
        self.minHeap = []
        self.maxHeap = []
        pass

    def addNum(self, num: int) -> None:
        if not self.minHeap and not self.maxHeap:
            heapq.heappush(self.minHeap, num)
            return

        if num < self.minHeap[0]:
            heapq.heappush(self.maxHeap, -num)
        else:
            heapq.heappush(self.minHeap, num)

        while abs(len(self.minHeap) - len(self.maxHeap)) > 1:
            if len(self.minHeap) > len(self.maxHeap):
                tmp = heapq.heappop(self.minHeap)
                heapq.heappush(self.maxHeap, -tmp)
            else:
                tmp = -heapq.heappop(self.maxHeap)
                heapq.heappush(self.minHeap, tmp)

    def findMedian(self) -> float:
        if len(self.minHeap) > 0 and len(self.minHeap) == len(self.maxHeap):
            return (self.minHeap[0] - self.maxHeap[0]) / 2
        if len(self.minHeap) > len(self.maxHeap):
            return self.minHeap[0]
        else:
            return -self.maxHeap[0]


if __name__ == "__main__":
    sol = MedianFinder()
    for i in (5, 3, 7, 2):
        sol.addNum(i)
        print(sol.findMedian())
    print("Running Solution...")

File: NeetCode/Heap/Median_Finder/Median_Finder.cpp
// Median_Finder.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class MedianFinder
{
private:
    priority_queue<int> maxHeap;
    priority_queue<int, vector<int>, greater<int>> minHeap;
    int count;

public:
    MedianFinder()
    {
        count = 0;
    }

    void addNum(int num)
    {
        count++;
        if (maxHeap.empty())
        {
            maxHeap.push(num);
        }
        else
        {
            if (maxHeap.top() > num)
            {
                maxHeap.push(num);
            }
            else
            {
                minHeap.push(num);
            }
        }

        if (maxHeap.size() > minHeap.size() + 1)
        {
            minHeap.push(maxHeap.top());
            maxHeap.pop();
        }

        if (maxHeap.size() < minHeap.size())
        {
            maxHeap.push(minHeap.top());
            minHeap.pop();
        }
    }

    double findMedian()
    {
        if (count % 2 == 1)
        {
            return maxHeap.top();
        }
        return (maxHeap.top() / 2.0) + (minHeap.top() / 2.0);
    }
};
